### ISSUE TITLE
[BUG] Harvester upgrade fails with witness node in cluster

### ISSUE THREAD

**Issue by SnakebiteEF2000** [harvester/harvester#6972]
**Describe the bug**
The upgrade failed from 1.3.1 to 1.3.2 on a 5 node harvester cluster.
2 physical master nodes
2 physical worker nodes
1 VM witness node

Tee upgrade get stuck at the step "upgrading system service"

After trying the trouble shooting steps in the docs this seems to be the problem:
instance-manager (aio) (image=longhornio/longhorn-instance-manager:v1.6.2) count is not 1 on node servername0000, will retry... 

I suspect the upgrade process somehow is not aware of the witness nodes inability to schedule this service.

**To Reproduce**
Start the upgrade from 1.3.1 to 1.3.2 with witness node in the cluster

**Expected behavior**
Successful upgrade

**Support bundle**
Sadly company policy does not allow the upload of the support bundle!

**Environment**
 - Harvester ISO version: 1.3.1
4 Poweredge R630 (within recommended specs)
1 vsphere based witness node
**Additional context**

**Comment by Vicente-Cheng** [comment]
Hi @SnakebiteEF2000, 

>After trying the trouble shooting steps in the docs this seems to be the problem:
instance-manager (aio) (image=longhornio/longhorn-instance-manager:v1.6.2) count is not 1 on node servername0000, will retry...

Could you check the instance-manager pod number? It should only have one instance-manager pod for each node.

**Comment by SnakebiteEF2000** [comment]
Yes it tries to startup 1 pod per node but the witness node does not start the pod as it is considered workload i guess. thats why the upgrade job just waits for the pod to be scheduled which will never happen.

**Comment by ibrokethecloud** [comment]
would it be possible for you to please generate a support bundle while the cluster is stuck in this state?

**Comment by bk201** [comment]
@SnakebiteEF2000 is `servername0000` the witness node and can you provide:
```
kubectl get instancemanager.longhorn.io -A
```

**Comment by bk201** [comment]
We can reproduce this with a 5-node cluster (2 mgmt, 1 witness, 2 workers).

**Comment by SnakebiteEF2000** [comment]
Hi @bk201,
Sadly I can't provide a support bundle due to company policy. The cluster also got rebuilt from scratch with the 1.3.2 image due to downtime and I am unable to retrieve new information from the cluster. Can I help in any other way like testing?

Edit: Yes `servername0000` is the witness node!

**Comment by Vicente-Cheng** [comment]
Hi @SnakebiteEF2000 
Here is the workaround:
TL;DR, patch the script and replace the image, then respawn the pod

1. find the pod running node
```
# kubectl get pods -A -o wide |grep apply-manifests
harvester-system                  hvst-upgrade-sdv8r-apply-manifests-bhqpz                                   1/1     Running     0               72m     10.52.2.4      harvester-node-2   <none>           <none>
```
2. goto harvester-node-2
3. find the running pod and copy the script out
```
# docker ps -a |grep upgrade_manifests.sh
578a5e2a64b3    docker.io/rancher/harvester-upgrade:v1.3.2                             "upgrade_manifests.sh"    About an hour ago    Up                  k8s://harvester-system/hvst-upgrade-sdv8r-apply-manifests-bhqpz/apply
# docker cp 578a5e2a64b3:/usr/local/bin/upgrade_manifests.sh .
# ls
bin  upgrade_manifests.sh
```
4. patch the upgrade_manifests.sh
```
# sed -i '/Wait for instance-manager/{n;s/kubectl get nodes/kubectl get nodes.longhorn.io -n longhorn-system/;}' upgrade_manifests.sh
```
5. copy back to the running pod
```
# docker cp upgrade_manifests.sh 578a5e2a64b3:/usr/local/bin/upgrade_manifests.sh
```
6. commit to the new image
```
# docker commit 578a5e2a64b3 tmp/image:dev
WARN[0000] Image lacks label "nerdctl/platform", assuming the platform to be "linux/amd64"
sha256:3d93c93d1f093434183909ebb598da17542a55b0f7950b6f5b94bb6fd86fb0d2
```
7. backup the old image and overwrite the image to the old one
```
# docker tag rancher/harvester-upgrade:v1.3.2 rancher/harvester-upgrade:backup
# docker tag tmp/image:dev rancher/harvester-upgrade:v1.3.2
```
8. delete the running apply-manifests pod, it will recreate and use the new image
```
# kubectl get pods -A |grep apply-manifest
harvester-system                  hvst-upgrade-sdv8r-apply-manifests-bhqpz                                   1/1     Running     0               91m
# kubectl delete pods hvst-upgrade-sdv8r-apply-manifests-bhqpz -n harvester-system
```

NOTE: The apply-manifests pod should respawn in the same node. If the apply-manifests pod respawns on a different node, please copy the patched image to the whole node and delete the pod again.

**Comment by albinsun** [comment]
Test PASS for both workaround and the fix in `v1.4-head`, close as fixed.


## 🟢 Verify Fix ( `v1.3.2` -> `v1.4-bede7c36-head`)
### Environment
* `harvester-v1.4-bede7c36-head`
  * Profile: 5 nodes  AMD64 QEMU/KVM with profile: **default**-**default**-**witness**-**workder**-**workder** 
  * ui-source: `Auto`
 
### Steps
1. Before Upgrade
     ![Image](https://github.com/user-attachments/assets/5b6b3dba-f664-43d6-ae63-6e87233c824c)
    VM & RKE2 cluster
    ![Image](https://github.com/user-attachments/assets/022816f1-1ec2-4e4d-b5f7-91080875afe3)

1. After Upgrade
   * Cluster
      ![Image](https://github.com/user-attachments/assets/6a2dc382-6f97-4080-ac91-331940db6dad)
   * VM a& RKE Cluster
      ![Image](https://github.com/user-attachments/assets/eaf2bd7f-58ae-4dc7-aa4d-bd50f961b26b)




## 🟢 Verify Workaround
### Environment
* <code>v1.3.2</code> -> <code>v1.4.0-rc5</code>
  * Profile: 4 nodes  AMD64 QEMU/KVM with profile: **default**-**default**-**witness**-**workder**
  * ui-source: `Auto`
 
### Reproduce
1. <details>
    <summary> Can repeoduce on <code>v1.3.2</code> -> <code>v1.4.0-rc5</code>
    </summary>
 
    ![Image](https://github.com/user-attachments/assets/3f7fa956-5d43-47fe-b975-e557877a845d)   
    </details>

### Try workaround mentioned in https://github.com/harvester/harvester/issues/6972#issuecomment-2472960181
1. find the pod running node (here **harvester-node-2**)
    ![Image](https://github.com/user-attachments/assets/6e6b62c1-ff97-45bb-ac3d-4197348206cb)
1. goto **harvester-node-2**
    ![Image](https://github.com/user-attachments/assets/5ebbd2b1-d05f-4300-acea-42c7fc03349c)  
1. find the running pod and copy the script out
    ![Image](https://github.com/user-attachments/assets/fb20c811-216f-432c-a5f7-ba00a5fbcbf7)
1. patch the upgrade_manifests.sh
    ![Image](https://github.com/user-attachments/assets/3ccfda1f-9379-4795-b92c-02787b6cd4a6)
1. copy back to the running pod
    ![Image](https://github.com/user-attachments/assets/da0d3d61-510a-40bd-af9e-350c8f18d001)
1. commit to the new image
    ![Image](https://github.com/user-attachments/assets/327f225b-7e80-47c3-b438-703792b351b3)
1. backup the old image and overwrite the image to the old one
    ![Image](https://github.com/user-attachments/assets/1b096d47-1a33-4cfe-bc38-1a5018b2140b)
1. delete the running apply-manifests pod, it will recreate and use the new image
    ![Image](https://github.com/user-attachments/assets/51105c2c-c0cb-476b-a497-447013f0e0a9)

### Upgrade goes on 
![Image](https://github.com/user-attachments/assets/40388e4a-f3d9-4c31-b842-de4d048992bf)
and success
![Image](https://github.com/user-attachments/assets/c04b3f4c-5155-4920-82b5-26a92a8a0a8b)
Can import to Rancher and create RKE2 cluster
![Image](https://github.com/user-attachments/assets/d8fb74d8-eac7-4f06-aa15-55a58ab84648)
